crates/extensions/c8y_firmware_manager/src/actor.rs

```diff
@@ -0,0 +1,772 @@
+        if let Err(err) =
+            // Addressing the new firmware operation to further step.
+            self
+                .handle_firmware_download_request_child_device(
+                    smartrest_request.clone(),
+                    op_id.as_str(),
+                )
+                .await
+        {
+            self.fail_operation_in_cloud(child_id, Some(&op_id), &err.to_string())
+                .await?;
+        }
+
+        Ok(())
+    }
+
+    // Check if the firmware file is already in cache.
+    // If yes, going to further step going to the last step; to publish a firmware request to child device.
+    // Otherwise, send a download request to the downloader.
+    async fn handle_firmware_download_request_child_device(
+        &mut self,
+        smartrest_request: SmartRestFirmwareRequest,
+        operation_id: &str,
+    ) -> Result<(), FirmwareManagementError> {
+        let firmware_url = smartrest_request.url.as_str();
+        let file_cache_key = digest(firmware_url);
+        let cache_file_path = self
+            .config
+            .validate_and_get_cache_dir_path()?
+            .join(&file_cache_key);
+
+        if cache_file_path.is_file() {
+            info!(
+                "Hit the file cache={}. File download is skipped.",
+                cache_file_path.display()
+            );
+            // Publish a firmware update request to child device.
+            self.handle_firmware_update_request_with_downloaded_file(
+                smartrest_request,
+                operation_id,
+                &cache_file_path,
+            )
+            .await?;
+        } else {
+            info!(
+                "Awaiting firmware download for op_id: {} from url: {}",
+                operation_id, firmware_url
+            );
+
+            let auth = if self
+                .config
+                .c8y_end_point
+                .url_is_in_my_tenant_domain(firmware_url)
+            {
+                if let Ok(token) = self.message_box.jwt_retriever.await_response(()).await? {
+                    Some(Auth::new_bearer(&token))
+                } else {
+                    return Err(FirmwareManagementError::NoJwtToken);
+                }
+            } else {
+                None
+            };
+
+            // Send a request to the Downloader to download the file asynchronously.
+            let download_request = DownloadRequest::new(firmware_url, &cache_file_path, auth);
+
+            self.message_box
+                .download_sender
+                .send((operation_id.to_string(), download_request))
+                .await?;
+            self.reqs_pending_download
+                .insert(operation_id.to_string(), smartrest_request);
+        }
+        Ok(())
+    }
+
+    // This function is called after download completes by downloader.
+    // If the download is successful, going to the last step; to publish a firmware request to child device.
+    // Otherwise, go
```

Comment:
```suggestion
    // This function is called on receiving a DownloadResult from the DownloaderActor or when the firmware file is already available in the cache.
    // If the download is successful, publish a firmware request to child device with it
    // Otherwise, fail the operation in the cloud
```